### ISSUE TITLE
beancount should consider warning when the _decimal C extension is missing

### ISSUE THREAD

**Issue by blais** [beancount/beancount#90]
**[Original report](https://bitbucket.org/blais/beancount/issue/90) by Jeremy Maitin-Shepard (Bitbucket: [jbms](https://bitbucket.org/jbms), GitHub: [jbms](https://github.com/jbms)).**

----------------------------------------

Python 3 includes a much faster C implementation of the decimal library, but this is not always installed by default.  For example, on Arch Linux it is only available if the mpdecimal package is installed, and I didn't happen to have that package installed on my computer.

This dramatically improves the performance of Beancount, and therefore it might be worth suggesting in the documentation to be sure it is available, or checking automatically in the code and issuing a warning if it is not found.  It can be tested by trying to:

import _decimal

This check might have to be updated eventually in a later version of Python, but since it would just be a warning it is not as big a deal if it breaks.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I'm well aware of this, cdecimal was required (and documented) for Beancount before Python 3.3. I can't imagine using the Python implementation, its' way too slow. In Python 3.2 and before it was an external install.

However, my understanding is that since Python 3.3 it's the default implementation and is part of Python's source code (just verified it's still the case in 3.4.3). Beancount already requires Python 3.3 for this purpose, so AFAIK nothing needs be mentioned. In fact, "import _decimal" fails in 3.4.3.

I'd venture to say that if you don't have it you probably have an incomplete Python installation.
Can you confirm that you're witnessing an Arch installation of 3.3 or above without cdecimal?

**Comment by blais** [comment]
**Original comment by Jeremy Maitin-Shepard (Bitbucket: [jbms](https://bitbucket.org/jbms), GitHub: [jbms](https://github.com/jbms)).**

----------------------------------------

I have Python 3.5.0 installed through the system package manager on Arch Linux.

As stated here in the Dependencies section: https://www.archlinux.org/packages/extra/x86_64/python/

the package seems to be set up to make the dependency on mpdecimal optional, with the decimal C extension being available only if mpdecimal is installed.  It was only yesterday, in the course of profiling beancount-web, that I realized it was missing.

On Python 3.5, import _decimal works.  I guess on 3.4 the naming isn't the same.

A better way to detect it is probably:

import types, decimal

isinstance(decimal.Decimal().sqrt, types.BuiltinFunctionType)

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Makes sense. I'll add a warning.
Thanks for reporting this.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

https://bitbucket.org/blais/beancount/commits/b3becc217a20d66b42b791adf418f572dde82695

**Comment by blais** [comment]
**Original comment by Jeremy Maitin-Shepard (Bitbucket: [jbms](https://bitbucket.org/jbms), GitHub: [jbms](https://github.com/jbms)).**

----------------------------------------

Thanks very much.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Wow. Just detected that one of the machines I'm using (an Ubuntu 14.04 install) suffers from the same problem.

**Comment by blais** [comment]
**Original comment by Adrian Ratnapala (Bitbucket: [adrianratnapala](https://bitbucket.org/adrianratnapala), GitHub: [adrianratnapala](https://github.com/adrianratnapala)).**

----------------------------------------

I am trying beancount (revision f3377eb58fe) with python 3.5.2 on archlinux.   I still get the warning until I explictly instaled "extra/mpdecimal".

The comment in `number.py` says "I believe this subtle problem should go away with Python-3.5 and above."   -- but apparently we still have it in python3.5

I think the warning should be changed to say "Consider installing cdecimal or mpdecimal".  Or some such.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

That sounds like an archlinux issue, i.e. python not depending on its full list of dependencies and the decimal wrapper handling the failure to load the dso.

Can you list the set of files installed by the "mpdecimal" archlinux package?

**Comment by blais** [comment]
**Original comment by Adrian Ratnapala (Bitbucket: [adrianratnapala](https://bitbucket.org/adrianratnapala), GitHub: [adrianratnapala](https://github.com/adrianratnapala)).**

----------------------------------------

I got beancount from the AUR, and indeed the PKGBUILD does not have `mpdecimal` listed as a dependency.  Is this the problem, or do you mean that some of the actual python configuration is wrong under Arch?

Anyway the file list from:

    pacman -Ql mpdecimal

is (apart from some documentation)

    mpdecimal /usr/
    mpdecimal /usr/include/
    mpdecimal /usr/include/mpdecimal.h
    mpdecimal /usr/lib/
    mpdecimal /usr/lib/libmpdec.so
    mpdecimal /usr/lib/libmpdec.so.2
    mpdecimal /usr/lib/libmpdec.so.2.4.2
    mpdecimal /usr/share/
    mpdecimal /usr/share/doc
     <SNIPPED>
    mpdecimal /usr/share/licenses
    /usr/share/licenses/mpdecimal/LICENSE.txt

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Thanks Adrian. 
This file list confirms what I thought: It's not the Python modules which are incomplete, rather, Python's cdecimal module is installed but it depends on the libmpdec.so being present. The official source code dispatches off of that being present, and if the dso cannot be loaded (because some of its runtime dependencies are absent), an ImportError will be raised and caught and fallback on the Python module. See Python's source code here:

https://hg.python.org/cpython/file/tip/Lib/decimal.py

If you'd like to confirm this, find the module in your Python installation:

bergamot:~$ locate --regex '/usr.*_decimal.*\.so'
/usr/lib/python3.5/lib-dynload/_decimal.cpython-35m-x86_64-linux-gnu.so

As above, it should resolve in a dso being correctly installed.

In other words, Python's C decimal implementation is installed correctly, but Arch does not correctly install its mpdecimal dependencies, i.e., it allows you to install Python without installing all of the dso's it depends on.

This is a problem with Arch; not Python, not Beancount.

**Comment by blais** [comment]
**Original comment by Adrian Ratnapala (Bitbucket: [adrianratnapala](https://bitbucket.org/adrianratnapala), GitHub: [adrianratnapala](https://github.com/adrianratnapala)).**

----------------------------------------

What I am trying to figure out is which aspect of Arch should gbe fixed.

It seems that "/usr/lib/python3.5/decimal.py" is owned directly by the package "python 3.5.2-1".  Does that mean the python PKGBUIL should pull-in (or include) mpdecimal?

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I don't know how they treat their dependencies. In my book, if a binary package has a runtime dependency on another dso it should always install that dso if you install the package. The Arch maintainers may disagree. No idea.

Note that Python's decimal module still works without mpdecimal-- it's just slow. I suspect that's the argument they may bring up for not enforcing the dependency and for everyone to install mpdecimal by default (even though it looks like a harmless C package).

In any case, I'll add a note to my Install document.
Thank you,

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

Issue is due to Arch linux and its borked Python installation.
